### PR TITLE
Update coding guidelines

### PR DESCRIPTION
**Pull request by daxian-dbw**
Update coding guidelines to make it more concrete and useful in a review process.

### CODE REVIEW

**Comment by daxian-dbw** [comment]
Please feel free to question the content I put in the doc as well as suggest more content.

File: .github/CONTRIBUTING.md
Comment (andschwa) [inline .github/CONTRIBUTING.md]: I like this.

File: docs/dev-process/coding-guidelines.md
Comment (andschwa) [inline docs/dev-process/coding-guidelines.md]: Maybe a note to explicitly avoid UTF-16, and a warning that some Windows editors may default to this.

File: docs/dev-process/coding-guidelines.md
Comment (andschwa) [inline docs/dev-process/coding-guidelines.md]: Maybe a note about avoiding exceptions as much as possible? Especially avoid using exceptions for logic (reserve them for unrecoverable errors such as those from the OS or libraries).

File: docs/dev-process/coding-guidelines.md
Comment (andschwa) [inline docs/dev-process/coding-guidelines.md]: Maybe mention some of the existing utility classes.

**Review by andyleejordan — APPROVED** [review]
Left some ideas, but looks great overall.

File: docs/dev-process/coding-guidelines.md
Comment (lzybkr) [reply docs/dev-process/coding-guidelines.md]: I'd like to disallow all `BOM` encodings - tests that need a `BOM` are better off generating the file anyway.

**Review by lzybkr — APPROVED** [review]

File: .github/CONTRIBUTING.md
Comment (SteveL-MSFT) [reply .github/CONTRIBUTING.md]: May be worth adding that the body of the PR is used to create the ChangeLog, so try to have the first sentence explain benefit to end user

File: docs/dev-process/coding-guidelines.md
Comment (SteveL-MSFT) [inline docs/dev-process/coding-guidelines.md]: Update/remove

File: .github/CONTRIBUTING.md
Comment (daxian-dbw) [reply .github/CONTRIBUTING.md]: Added it to the doc.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Fixed -- Mention that all `bom` encoding should be avoided.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Good point. Added.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Add an example utility class `System.Management.Automation.StringToBase64Converter`

**Review by SteveL-MSFT — APPROVED** [review]

File: docs/dev-process/coding-guidelines.md
Comment (PaulHigin) [inline docs/dev-process/coding-guidelines.md]: I feel we should say a little more about what security risks are and should at least mention input validation, code injection, privilege escalation, data privacy, and DOS concerns.  We should have a separate document that goes into more detail but that can be a separate effort.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: `input validation` - this could be easily confused with the input validation within an API. Could you please elaborate a bit on it in the security context?

`data privacy` - do you mean data privacy breach? Can you please give an example?

> We should have a separate document that goes into more detail

That would be very useful. Can you please work on such a doc? If you are busy on things and cannot get to it in near future, I'm happy to work on a draft if you can provide me some resources and pointers.

Comment (iSazonov) [reply docs/dev-process/coding-guidelines.md]: Has CoreFX repo such document?

File: docs/dev-process/coding-guidelines.md
Comment (iSazonov) [inline docs/dev-process/coding-guidelines.md]: Prefer over LINQ?

File: docs/dev-process/coding-guidelines.md
Comment (vors) [inline docs/dev-process/coding-guidelines.md]: > `Verb-Object` pair such as **`LoadModule`**.

shoud it be `VerbObject` then?

File: docs/dev-process/coding-guidelines.md
Comment (vors) [inline docs/dev-process/coding-guidelines.md]: this is generally a linter job to enforce such things.
We have a hand-written linter for tabs vs spaces and such.
Do we consider to add a proper dotnet linter to CI and build.psm1?

File: docs/dev-process/coding-guidelines.md
Comment (iSazonov) [inline docs/dev-process/coding-guidelines.md]: I remember a recommendation to use static RegEx to benefit caching. Is this relevant? Maybe still mention compile RegEx and using StringBuilder if we have more 3-4 re-allocations.

**Review by iSazonov — COMMENTED** [review]
Sorry for many comments - most is minor - feel free to ignore.

File: docs/dev-process/coding-guidelines.md
Comment (vors) [inline docs/dev-process/coding-guidelines.md]: I think phrasing can be better here. It may be perceived as encouragement for creating a historical log in the comments, i.e.

```
// This line used to be bla-bla 
//  and was changed to bla-bla because previously we had a problem in bla-bla corner case.
```

File: docs/dev-process/coding-guidelines.md
Comment (PaulHigin) [reply docs/dev-process/coding-guidelines.md]: Input validation for public API or Cmdlet is exactly what I mean :).   That is where code injection can occur and I feel being aware of it should be part of all code reviews.  Likewise data privacy is being aware that private data must always be protected.  It is not something we have to be too concerned with with PowerShell (other than credentials and transferring remote data) but again something to be aware of during a code review.

I think we should just mention basic security areas to be aware of during a code review in a general way here, and then provide a link to a separate document where we can provide more details and examples.  I am happy to work on such a document but I am not sure when I can get to it.  I'll create an issue to track the work.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Here are my experiences:

1. Depending on how bad the reformatting affects the review, I will choose to either tolerate the formatting changes or block the PR and ask the author to revert the formatting changes.
    * For example, some PRs has changes that remove the trailing spaces. When that kind of formatting changes doesn't affect review much, I will tolerate them.
    * But for some other PRs, the formatting changes might cause the review to be difficult, and I will block the PR and ask the author to revert formatting changes.

2. We ask authors to revert the formatting changes and suggest them to submit a separate PR, but they are not required to submit a PR to fix the formatting. You can give them the choices as for how to revert the reformatting changes -- submit style PR and merge and then rebase the original PR is certainly one way to do it; or they can just revert the formatting changes and submit a style PR later.

3. I don't like tracking issue in this situation.

File: docs/dev-process/coding-guidelines.md
Comment (PaulHigin) [reply docs/dev-process/coding-guidelines.md]: #4759

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: The layout conventions are mostly from https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

I think we shouldn't go that details about how to use empty lines.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Will remove `unnecessary`.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Yes, `sorted alphabetically` is too much detailed. I will remove it.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Static and const fields are fields too, right?
For fields comes with properties, yes, we should add exceptions.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: It's used a lot in [Compiler.cs](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/Compiler.cs#L60)

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Yes, reviewer and maintainer should pay attention to public APIs and let the committee review public APIs.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: What do you mean here?

File: docs/dev-process/coding-guidelines.md
Comment (lzybkr) [reply docs/dev-process/coding-guidelines.md]: The motivation here is to easily and more accurately find references - simple strings aren't not always a reference.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: I will change the above item to be
Make sure the added/updated comments are **useful**, accurate and easy to understand.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: I think StringBuilder is worth mentioning -- string operations are quite common. `RegEx` is too specific to me.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: I don't think that belongs to coding best practices. And `use CoreFX APIs` is mentioned in the portable code section:
```
Going forward, we try to depend on .NET Core to handle platform differences,
so avoid adding new P/Invoke calls where a suitable alternative exists in .NET Core.
```

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: hard-code -- `fix (data or parameters) in a program in such a way that they cannot be altered without modifying the program.`  See https://www.google.com/search?source=hp&q=hard-coding&oq=hard-coding&gs_l=psy-ab.3..0i67k1j0i20k1j0i30k1l2.219.1708.0.1826.12.12.0.0.0.0.106.959.11j1.12.0....0...1.1.64.psy-ab..0.12.957.6..0j35i39k1j0i131k1j0i10k1.NyUenydrd2Q

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Will add it.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Will mention this in this section.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Will fix.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: @TravisEz13 has some ideas to enforce the formatting check in CI.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: I also added `Begin comment text with an uppercase letter. It's recommended to end comment text with a period but not required.`

I don't think we should leave a review comment about the ending period unless it's causing confusion, for example, there are multiple sentences in comments, and it's hard to tell where a sentence ends without the ending period.

File: docs/dev-process/coding-guidelines.md
Comment (iSazonov) [reply docs/dev-process/coding-guidelines.md]: I unlike tracking issue too. But without it we can forget and lost. So we should open style PR before continue original PR.

File: docs/dev-process/coding-guidelines.md
Comment (iSazonov) [reply docs/dev-process/coding-guidelines.md]: Closed.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Mentioned in `### Layout Conventions` as it's related to the code layout.

File: docs/dev-process/coding-guidelines.md
Comment (iSazonov) [reply docs/dev-process/coding-guidelines.md]: Thanks for clarify.

File: docs/dev-process/coding-guidelines.md
Comment (iSazonov) [reply docs/dev-process/coding-guidelines.md]: Sorry, will this added in the doc?

File: .github/CONTRIBUTING.md
Comment (daxian-dbw) [reply .github/CONTRIBUTING.md]: Words adjusted.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: The sentences are reworded to:
```
* Add comments where the code is not trivial or could be confusing.
* Add comments where a reviewer needs help to understand the code.
```

File: docs/dev-process/coding-guidelines.md
Comment (iSazonov) [reply docs/dev-process/coding-guidelines.md]: What is a context here? "We prefer use 'for' instead of LINQ"?

File: docs/dev-process/coding-guidelines.md
Comment (iSazonov) [reply docs/dev-process/coding-guidelines.md]: Closed.

File: docs/dev-process/coding-guidelines.md
Comment (iSazonov) [reply docs/dev-process/coding-guidelines.md]: Thanks for clarify!

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Not in this doc as it's more of guideline or best practices for maintainers. I think a guideline (best practices) for maintainer would be useful.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: I don't think we should make it a requirement for the author to submit a style-only PR in this situation. It's acceptable if they just revert the formatting changes, and it would be a plus if they are willing to go one step further to fix the style/formatting as well.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Yes, I think it's prefer looping over collections directly over LINQ.
I will try to clarify.

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: Add the following:
```
... may introduce security risks,
such as code injection caused by the lack of input validation,
privilege escalation due to the misuse of impersonation,
or data privacy breach with a plain text password.
```

**Comment by daxian-dbw** [comment]
All comments are addressed. Please take another look, thanks!

**Review by iSazonov — APPROVED** [review]
Thanks!

**Comment by daxian-dbw** [comment]
@PaulHigin your comment has been addressed. Can you please take another look?

File: docs/dev-process/coding-guidelines.md
Comment (TravisEz13) [reply docs/dev-process/coding-guidelines.md]: Adding the motivation may help clarify the guidance.

**Review by TravisEz13 — APPROVED** [review]

**Review by PaulHigin — APPROVED** [review]
LGTM

File: docs/dev-process/coding-guidelines.md
Comment (daxian-dbw) [reply docs/dev-process/coding-guidelines.md]: @TravisEz13 the motivation is added.

**Comment by iSazonov** [comment]
@daxian-dbw Is the PR ready to merge? (Squash and merge?)

**Comment by daxian-dbw** [comment]
@iSazonov Yes, I think it's ready to go.

**Comment by iSazonov** [comment]
@daxian-dbw Many thanks!

**Pull request merged**
